[BEAM-1017] Add RedisIO.read()

jbonofre opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

jbonofre commented:
R: @jkff

jkff commented on sdks/java/io/redis/pom.xml:
```diff
@@ -0,0 +1,93 @@
+
+  <artifactId>beam-sdks-java-io-redis</artifactId>
+  <name>Apache Beam :: SDKs :: Java :: IO :: Redis</name>
+  <description>IO to send and receive messages from Redis. Redis is advanced key-value
```
Don't think you need to describe what is Redis in the documentation of a connector to Redis - if someone decided to use this connector, they probably already know what Redis is.

Also, in this PR the IO provides only the Read transform.

Also, I didn't realize that you're implementing just the pubsub thing. Is this the only read-like functionality Redis provides? If not, it should be named differently than "Read".

jbonofre replied:
Actually, in the original PR, the RedisIO provided both RedisIO for the key store and RedisPubSubIO for the pub/sub. For this PR, I only included the key-value store part.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+import redis.clients.jedis.JedisCommands;
+
+/**
+ * {@code RedisConnection} describes and wraps a connection to Redis server or cluster.
```
It's not a connection, it's the configuration necessary to establish a connection - please rename the class to RedisConnectionConfiguration or something like that.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+    return builder().setTimeout(timeout).build();
+  }
+
+  public Jedis connect() {
```
Document this. Among other things, explain that it uses the first node in the list; by the way, is it allowed to call this method if there's more than 1 node? When should one call this method vs. connectCluster?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+@AutoValue
+public abstract class RedisConnection implements Serializable {
+
+  @Nullable abstract List<String> nodes();
```
Shouldn't this be a `List<HostAndPort>`, with parsing happening somewhere else?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+
+  protected Properties info(Jedis jedis) throws Exception {
+    String info = jedis.info();
+    jedis.info("cluster_enabled");
```
What does this do?

jbonofre replied:
It checks if we connect to a cluster or a single instance.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+  protected boolean isClusterEnabled(Jedis jedis) throws Exception {
+    Properties properties = info(jedis);
+    String value = properties.getProperty("cluster_enabled");
+    if (value != null && !value.equals("0")) {
```
Reduce to a single return statement.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+    return false;
+  }
+
+  public void populateDisplayData(DisplayData.Builder builder) {
```
`@Override`?

jbonofre replied:
This `populateDisplayData()` doesn't override (it's in the connection configuration, not in the `PTransform`).

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnection.java:
```diff
@@ -0,0 +1,124 @@
+
+  public void populateDisplayData(DisplayData.Builder builder) {
+    StringBuilder stringBuilder = new StringBuilder();
+    for (String node : nodes()) {
```
Joiner.on(" ").join(nodes())

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+    /**
+     * The Redis commands related to read of key-value pairs.
+     */
+    public enum Command {
```
What's the point of having this enum and the relevant getter/setter methods, if it only has a single value?

jbonofre replied:
The plan is to add other commands support in the future (in another PR).

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+    @Nullable abstract RedisConnection connection();
+    @Nullable abstract Command command();
+    @Nullable abstract String keyPattern();
+    @Nullable abstract RedisService redisService();
```
Is connection() redundant with redisService()? Both specify how to connect to Redis, right?

jbonofre replied:
Yes, `RedisConnection` could be part of the `RedisService`.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+    /**
+     * Define the connection to the Redis server.
+     *
+     * @param connection The {@link RedisConnection}.
```
Remove useless parts from all comments (those that don't add any information that isn't already obvious from the function signature).

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    /**
+     * Define the Redis service to use.
```
This comment didn't help me understand what this method is for and how/when to use it.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(connection() != null || redisService() != null, "RedisIO.read() requires a "
```
This reinforces my impression that these two should perhaps be merged.

jbonofre replied:
Agree.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+      checkState(connection() != null || redisService() != null, "RedisIO.read() requires a "
+          + "connection to be set withConnection(connection) or a service to be set "
+          + "withRedisService(service)");
+      checkState(command() != null,  "RedisIO.read() requires a command to be set via "
```
It's weird to force users to do something about a parameter that has only one legal value. Just delete the enum and the getters/setters from the API.

jbonofre replied:
I would prefer to avoid that as new commands will come in another PR.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+  /**
+   * A bounded source reading key-value pairs from a Redis server.
+   */
+  // visible for testing
```
Use the appropriate annotation instead of a comment
`@VisibleForTesting`

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+    @Override
+    public List<RedisSource> splitIntoBundles(long desiredBundleSizeBytes,
+                                              PipelineOptions pipelineOptions) {
+      // TODO cluster with one source per slot
```
I don't understand this comment.

jbonofre replied:
In the case of a cluster connection, it would be possible to split the read with one source per node in the cluster.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,294 @@
+
+    @Override
+    public Coder<KV<String, String>> getDefaultOutputCoder() {
+      return KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of());
```
Are Redis keys and values always UTF-8 strings?

jbonofre replied:
Yes, it's a `String`.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisService.java:
```diff
@@ -0,0 +1,71 @@
+interface RedisService extends Serializable {
+
+  /**
+   * The interface of a class that can read from Redis.
```
Why not just use BoundedReader interface?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisService.java:
```diff
@@ -0,0 +1,71 @@
+  Reader createReader(String keyPattern) throws IOException;
+
+  /**
+   * Return an estimation of the size that could be read.
```
This comment is trivial.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java:
```diff
@@ -0,0 +1,120 @@
+    @Override
+    public boolean start() {
+      jedis = connection.connect();
+      Set<String> keys = jedis.keys(keyPattern);
```
The Jedis documentation specifically prohibits using this method.

http://javadox.com/redis.clients/jedis/2.2.0/redis/clients/jedis/Jedis.html#keys(java.lang.String)

Is there a better API you can provide?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java:
```diff
@@ -0,0 +1,120 @@
+    long samplingAverage = samplingSize / 10;
+    // db size
+    long dbSize = jedis.dbSize();
+    jedis.quit();
```
This will leak if anything above throws.

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,186 @@
+        });
+    long estimatedSizeBytes = source.getEstimatedSizeBytes(pipelineOptions);
+    LOG.info("Estimated size: {}", estimatedSizeBytes);
+    assertEquals(100, estimatedSizeBytes);
```
Explain why should it be 100?

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,186 @@
+
+    PAssert.thatSingleton(output.apply("Count", Count.<KV<String, String>>globally()))
+        .isEqualTo(1000L);
+    ArrayList<KV<String, String>> expected = new ArrayList<>();
```
You can put this at the top of the test, and at line 78 instead iterate over `expected` to avoid effectively duplicating the code describing the expected data.

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,186 @@
+    @Override
+    public long getEstimatedSizeBytes() {
+      long size = 0L;
+      Iterator<String> keysIterator = store.keySet().iterator();
```
You can iterate over it using a regular foreach loop. Also why do you need to look only at the 10 first elements in this mock implementation?

jbonofre commented:
Spark Redis connector is pretty close to what I did. For instance:

```
val stringRDD = sc.fromRedisKV("keyPattern*")
```

does a full scan on the partition. Let me improve that part.

jbonofre commented:
@jkff I addressed most of the comment and started the split implementation based on cluster slots. It requires some improvements, but the basic idea is already there.

jbonofre commented:
@jkff I think it's ready for a new round.

I have a question for you: as the splitting for Redis is based only on cluster mode with the number of slots, do you think it's worth to do any estimate size bytes ? I know that the runner can potentially use the estimated size, but from a RedisIO perspective, it's useless.

jbonofre commented:
@jkff what do you think about using apache/beam pull request (for visibility) and move forward step by step using commits. The first comment could propose the big steps, like for instance:
1. RedisIO.read() including tests and service
2. RedisIO.write() including tests and service
3. RedisIOIT
The review is incremental.

Thought ?

I'm a bit concern about the pull request not at Apache (especially for visibility).

jkff commented:
As usual, my concern is about exposing users to an insufficiently tested IO. IOs are hard to get right, and if it's insufficiently tested, it practically always means it has data loss bugs - I think this is a more important concern than PR visibility.

What defines "sufficiently tested" is a separate question of course. I would be happy enough with testing against an in-memory embedded Redis instance (if such a thing exists), like we do for JdbcIO, but I would _not_ be happy enough with testing against a mock Redis that we implemented ourselves - in that case, I'd be fine exposing it to users only after adding an IT _and_ having the IT be continuously running on Beam infrastructure.

It might make sense to discuss this on the mailing list if you feel differently about what the standard for exposing an IO to users should be.

jbonofre commented:
Maybe you get me wrong: I don't mean to merge on the main repo, I mean keep the pull request on "Apache" github, and move forward in this pull request with commits. I fully agree with you about tested and "evaluated" IOs. The three steps that I mentioned in my previous comment are still in the PR (not merge). The PR is merged at the end, when the three steps have been approved.

jbonofre commented:
@jkff I rebased and updated this pull request with:
- some fixes on the `RedisSource`
- use of a embedded Redis server for concrete tests
Do you think it's worth to keep `RedisService` ?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,168 @@
+    return builder().setPort(port).build();
+  }
+
+  public RedisConnectionConfiguration withAuth(String auth) {
```
What's an auth? It seems like auth() is never used.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,168 @@
+  }
+
+  public RedisConnectionConfiguration withTimeout(int timeout) {
+    return builder().setTimeout(timeout).build();
```
Check that timeout is non-negative? Document what does it mean if it's zero?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,168 @@
+    builder.addIfNotNull(DisplayData.item("timeout", timeout()));
+  }
+
+  public List<RedisService.RedisNode> getNodes() throws Exception {
```
This doesn't seem to throw Exception from any of the called methods

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,168 @@
+    try (Jedis jedis = connect()) {
+      List<Object> slots = jedis.clusterSlots();
+
+      for (Object slotInfoObj : slots) {
```
Link to documentation describing the format of clusterSlots()?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,168 @@
+      for (Object slotInfoObj : slots) {
+        List<Object> slotInfo = (List<Object>) slotInfoObj;
+
+        int startSlot = Integer.parseInt(slotInfo.get(0).toString());
```
What type is slotInfo.get(0) actually? Seems like it represents an integer, but is a different type - is it a `String`? In that case a type cast to `String` may be more explicit

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,248 @@
+        int slot = serviceFactory.apply(pipelineOptions).getKeySlot(keyPattern);
+
+        for (RedisService.RedisNode node : nodes) {
+          if (node.startSlot < slot && node.endSlot > slot) {
```
Is this correct? Shouldn't it be something like `node.startSlot <= slot && slot < node.endSlot`?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,248 @@
+
+    /**
+     * The estimate size bytes is based on sampling, computing average size of 10 random
+     * key/value pairs. This sampling average size is used with the Redis dbSize to get an
```
Note that this ignores key pattern.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisService.java:
```diff
@@ -0,0 +1,72 @@
+    public String host;
+    public int port;
+    public int timeout;
+    public int startSlot;
```
Document the members? E.g. startSlot/endSlot - is it inclusive/exclusive. And what's size. Etc.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+      Pipeline pipeline = jedis.pipelined();
+      Response<Set<String>> keysResponse;
+      if (source.keyPattern == null) {
+        keysResponse = pipeline.keys("*");
```
Could handle this by making source.keyPattern be `*` by default, and always non-nullable.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+      }
+      pipeline.syncAndReturnAll();
+
+      Set<String> keys = keysResponse.get();
```
This is loading all keys in memory, and moreover probably assuming that Redis can return them all in one response. Seems unlikely to work for a large database. Is this intentional / required - does Redis provide some kind of cursor/scan API?

jbonofre replied:
Good point. Let me refactore to use Redis scan command.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+    public boolean advance() {
+      if (keysIterator.hasNext()) {
+        String key = keysIterator.next();
+        String value = jedis.get(key);
```
This is doing 1 Redis RPC for every key - seems inefficient. Is there a way to get both keys and values, for keys matching a pattern?

jbonofre replied:
I don't think so without hash. Let me double check.

jbonofre replied:
I think I can use a Redis pipeline. Let me try.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+   * estimation of the actual database size.
+   */
+  @Override
+  public long getEstimatedSizeBytes() {
```
This seems likely to be very inaccurate, because it ignores the key pattern and it ignores the fact that the database may be storing many different kinds of keys, with wildly different value size distributions.

Rather than asking to make this more accurate, I'm asking to reconsider whether perhaps it's better to implement RedisIO.Read as a DoFn, and implement size estimation when somebody asks for it. I think, given how hard it is to make it accurate and how little it matters in practice, it's likely to be not worth the complexity.

jbonofre replied:
I agree. The only way to take key pattern in consideration would be to not use db info and actually retrieve the key/value pairs. Which is inefficient as well.

Would you like I refactore the IO as a `DoFn` ? In that case, it means I have to find a way to use the splitting I did (using the cluster slots).

jkff replied:
Yes, I think this should be refactored as a DoFn. For splitting, you need a traditional ParDo/reshuffle/ParDo sequence, where the first ParDo gets the primary node in each slot range (or the set of all replicas for that slot), and the second ParDo reads the key pattern from that node (falling back onto other nodes - I suppose Jedis can do that for you automatically?)

jbonofre commented:
Resuming my work on this one.

jbonofre commented:
@jkff I refactored to use scan + pipeline to optimize both the memory and the number of interaction with the Redis server (it's a first refactored version, I think that I can cleanup/optimize a bit the code to  be more elegant). The only pending point is about the estimate size. Let me check if I can refactore the estimated size to use the key pattern. If you think it's better to use a `DoFn`, please let me know I  will refactore accordingly.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,194 @@
+        int startSlot = Integer.parseInt((String) slotInfo.get(0));
+        int endSlot = Integer.parseInt((String) slotInfo.get(1));
+
+        if (slotInfo.size() <= MASTER_NODE_INDEX) {
```
This condition seems unnecessary - the loop below will just immediately exit if there's no slots.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,194 @@
+
+        // host infos
+        int size = slotInfo.size();
+        for (int i = MASTER_NODE_INDEX; i < size; i++) {
```
Probably more readable to just say "2" because that makes it more obvious that you're reading elements of slotInfo in order: 0, 1, 2, ...

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,194 @@
+        int size = slotInfo.size();
+        for (int i = MASTER_NODE_INDEX; i < size; i++) {
+          List<Object> hostInfos = (List<Object>) slotInfo.get(i);
+          if (hostInfos.size() <= 0) {
```
You actually need it to have size exactly 2, right? And if it's not 2, then you'd probably want to throw an exception rather than silently skip it?

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,194 @@
+    builder.addIfNotNull(DisplayData.item("timeout", timeout()));
+  }
+
+  public List<RedisService.RedisNode> getNodes() {
```
Seems like this method is never used

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,194 @@
+    return nodes;
+  }
+
+  private final List<RedisService.RedisNode> nodes = new ArrayList<>();
```
This can be a local variable in getClusterNodes

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,257 @@
+    /**
+     * Define the connectionConfiguration to the Redis server.
+     *
+     * @param connection The {@link RedisConnectionConfiguration}.
```
These `@param` and `@return` javadocs provide no information on top of what is already obvious from the function signature; likewise below.

jkff replied:
Comment still stands.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,257 @@
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      if (connectionConfiguration() != null) {
```
It can't be null

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,257 @@
+        List<RedisService.RedisNode> nodes = serviceFactory.apply(pipelineOptions)
+            .getClusterNodes();
+
+        int slot = serviceFactory.apply(pipelineOptions).getKeySlot(keyPattern);
```
This is incorrect - what if the key pattern is `*`? You'll get the single slot into which the key `*` would have mapped, but `*` is not a key, it's a pattern. Instead in this case you should have used all slots. I believe you always need to use all slots, because for any pattern there will be keys in all slots matching the pattern, because sharding of a Redis cluster is hash-based rather than range-based.

I suppose this will lead to some refactoring of what you need in RedisNode.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,257 @@
+
+        for (RedisService.RedisNode node : nodes) {
+          if (node.startSlot <= slot && slot < node.endSlot) {
+            redisSources.add(new RedisSource(keyPattern, serviceFactory, node));
```
This is not correct either, you'll get all replicas containing the given slot, and get all data duplicated respectively this many times. Instead, you need to get only one replica for each slot range.

jbonofre commented:
I rebased and now I'm starting the refactoring as a `DoFn`.

jbonofre commented:
@jkff I refactored `RedisIO` to use a `DoFn` and implement the `ReadAll` pattern. Can you take a look please ? Thanks !

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,135 @@
+   * Define the host name of the Redis server.
+   */
+  public RedisConnectionConfiguration withHost(String host) {
+    checkArgument(host != null, "RedisConnectionConfiguration.create().withHost(host) called "
```
Please update the validation messages to the new recommended format https://beam.apache.org/contribute/ptransform-style-guide/#validation

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,242 @@
+ *      RedisConnectionConfiguration.create().withHost("localhost").withPort(6379))
+ *    .withKeyPattern("foo*")
+ *
+ * }</pre>
```
Mention ReadAll too

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {
```
This should be in `expand` per new guidance.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,242 @@
+              new ReadFn(connectionConfiguration())
+          ))
+          .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+      // reparallelized ?
```
Yes please

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,242 @@
+      String cursor = ScanParams.SCAN_POINTER_START;
+      boolean finished = false;
+      while (!finished) {
+        ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
```
I forget: does this scan need to be closed?

jbonofre replied:
No need, `ScanResult` doesn't have to be close (and it's an unsupported operation).

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,171 @@
+  public void testGetRead() throws Exception {
+    try (EmbeddedRedis embeddedRedis = new EmbeddedRedis()) {
+      ArrayList<KV<String, String>> expected = new ArrayList<>();
+      Jedis jedis = new Jedis("localhost", embeddedRedis.getPort());
```
Will this work with IPV6?

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,171 @@
+      ArrayList<KV<String, String>> expected = new ArrayList<>();
+      Jedis jedis = new Jedis("localhost", embeddedRedis.getPort());
+      for (int i = 0; i < 1000; i++) {
+        KV<String, String> kv = KV.of("Key " + i, "Value " + i);
```
Put insertion of test data into a private utility method? Could also return a ConnectionConfiguration to use at line 57.

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,171 @@
+                  .withHost("localhost")
+                  .withPort(embeddedRedis.getPort())));
+
+      PAssert.thatSingleton(readAll.apply("Count", Count.<KV<String, String>>globally()))
```
This is unnecessary - containsInAnyOrder will already take care of this.

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,171 @@
+            .withConnectionConfiguration(RedisConnectionConfiguration.create()
+              .withHost("localhost")
+              .withPort(embeddedRedis.getPort()))
+            .withKeyPattern("Key 0"));
```
Would be good if the test verified that we're matching a pattern rather than a single key - specify a pattern that matches multiple keys? And, for good measure, also have a test where the pattern matches no keys?

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  @Test
+  public void testConnectionInfiniteTimeout() throws Exception {
```
What is this testing?

jbonofre commented:
@jkff I think I addressed your comments and did the changes (reparallelize, javadoc, ...). Can you take a new round on this one ? Thanks !

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,131 @@
+    abstract RedisConnectionConfiguration build();
+  }
+
+  public static RedisConnectionConfiguration create() {
```
Might make sense to have a create(host, port) method for convenience also.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,131 @@
+   */
+  public void populateDisplayData(DisplayData.Builder builder) {
+    builder.addIfNotNull(DisplayData.item("host", host()));
+    builder.addIfNotNull(DisplayData.item("port", port()));
```
All of these can not be null, so you can use add().

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisConnectionConfiguration.java:
```diff
@@ -0,0 +1,131 @@
+
+  private static final int MASTER_NODE_INDEX = 2;
+
+  @Nullable abstract String host();
```
It's not actually nullable - create() creates it non-null, and withHost() doesn't accept null.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,285 @@
+ *
+ * <h3>Reading Redis key/value pairs</h3>
+ *
+ * <p>RedisIO.Read provides a source which returns a bounded {@link PCollection} containing
```
Better refer to the factory method - `{@link #read}`. Likewise below, `{@link #readAll}`.

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,285 @@
+          "withConnectionConfiguration() is required");
+
+      return input
+          .apply(ParDo.of(
```
Icky formatting, can fit on one line

jkff commented on sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java:
```diff
@@ -0,0 +1,285 @@
+      extends PTransform<PCollection<KV<String, String>>, PCollection<KV<String, String>>> {
+
+    @Override public PCollection<KV<String, String>> expand(PCollection<KV<String, String>> input) {
+      // breaking fusion
```
I suggest to explicitly say here that this is copied from JdbcIO with the same caveats, because this code is making some subtle assumptions and it's good to make it traceable to the source of the assumptions.

Also, consider rebasing on https://github.com/apache/beam/pull/3890 and using Reshuffle.viaRandomKey() when that goes in.

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,163 @@
+      insertKV(embeddedRedis);
+
+      ArrayList<KV<String, String>> expected = new ArrayList<>();
+      for (int i = 0; i < 1000; i++) {
```
Maybe use a smaller number so the test runs faster?

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,163 @@
+
+  @Test
+  public void testGetRead() throws Exception {
+    try (EmbeddedRedis embeddedRedis = new EmbeddedRedis()) {
```
All these tests create and tear down an EmbeddedRedis, and call insertKV. This begs for a `@Before`/`@After` method, to deduplicate the code.

jkff commented on sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java:
```diff
@@ -0,0 +1,163 @@
+      PCollection<KV<String, String>> read = pipeline.apply(
+          RedisIO.read()
+            .withConnectionConfiguration(RedisConnectionConfiguration.create()
+              .withHost("localhost")
```
This is the same as createConnection(embeddedRedis). I'd suggest to put it in a member variable.

jbonofre commented:
Awesome, thanks, I'm addressing your comments, squash and merge. Then I will start the write part.
